File: Cpp/Client/SimpleClient.cpp
#include <iostream>
#include <boost/asio.hpp>

using boost::asio::ip::tcp;

int main() {
    try {
        boost::asio::io_context io_context;
        tcp::socket socket(io_context);
        socket.connect(tcp::endpoint(boost::asio::ip::address::from_string("127.0.0.1"), 8080));

        std::string message_body = "Hello from C++ Boost.Asio!";
        uint32_t message_length = message_body.size();

        // Send the message length followed by the message body
        boost::asio::write(socket, boost::asio::buffer(&message_length, sizeof(message_length)));
        boost::asio::write(socket, boost::asio::buffer(message_body));

        // Read the response length
        uint32_t response_length;
        boost::asio::read(socket, boost::asio::buffer(&response_length, sizeof(response_length)));

        // Read the response body
        std::vector<char> response_body(response_length);
        boost::asio::read(socket, boost::asio::buffer(response_body));

        std::cout << "Reply from server: " << std::string(response_body.begin(), response_body.end()) << std::endl;

    } catch (std::exception& e) {
        std::cerr << e.what() << std::endl;
    }

    return 0;
}

File: Rust/Client/simple_client.rs
use tokio::net::TcpStream;
use tokio::io::{AsyncReadExt, AsyncWriteExt};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Connect to the server
    let mut stream = TcpStream::connect("127.0.0.1:8080").await?;

    // Message to send
    let message_body = "Hello from Rust client!";
    let message_length = message_body.len() as u32;

    // Send the message length followed by the message body
    stream.write_all(&message_length.to_le_bytes()).await?;
    stream.write_all(message_body.as_bytes()).await?;

    // Read the response length
    let mut response_length_buf = [0; 4];
    stream.read_exact(&mut response_length_buf).await?;
    let response_length = u32::from_le_bytes(response_length_buf);

    // Read the response body
    let mut response_body = vec![0; response_length as usize];
    stream.read_exact(&mut response_body).await?;

    // Convert response body to string
    let response_str = String::from_utf8(response_body)?;

    println!("Reply from server: {}", response_str);

    Ok(())
}
